src/backend/distributed/planner/local_plan_cache.c

```diff
@@ -99,6 +96,66 @@ CacheLocalPlanForShardQuery(Task *task, DistributedPlan *originalDistributedPlan
 }
 
 
+/*
+ * GetLocalShardQuery is a helper function which generates the local shard
+ * query based on the jobQuery.
+ *
+ * It is not guaranteed to have consistent attribute numbers on the shards
+ * and on the shell (e.g., distributed/reference tables) due to DROP COLUMN
+ * commands. To avoid any edge cases due to such discrepancies, we first
+ * deparse the jobQuery with the tables replaced to shards, and parse the
+ * query string back. This is normally a very expensive operation, however
+ * we only do it once per cached local plan, which is acceptable.
+ */
+static Query *
+GetLocalShardQuery(Query *jobQuery, Task *task, ParamListInfo orig_paramListInfo)
+{
+	StringInfo localShardSqlQuery = makeStringInfo();
+	if (jobQuery->commandType == CMD_INSERT)
+	{
+		/*
+		 * For INSERT queries we cannot use pg_get_query_def. Mainly because we
+		 * cannot run UpdateRelationToShardNames on an INSERT query. This is
+		 * because the PG deparsing logic fails when trying to insert into a
+		 * RTE_FUNCTION (which is what will happen if you call
+		 * UpdateRelationToShardNames).
+		 */
+		deparse_shard_query(jobQuery, task->anchorDistributedTableId, task->anchorShardId,
+							localShardSqlQuery);
+	}
+	else
+	{
+		UpdateRelationToShardNames((Node *) jobQuery, task->relationShardList);
+		pg_get_query_def(jobQuery, localShardSqlQuery);
+	}
```

Comment:
Is there a reason why we cannot use TaskQueryString here?

Reply:
One minor difference is that I used to free the generated SQL. But, when we use `TaskQueryString ` it sets the string to a pointer in the task, so we cannot free.

I think we should prefer having less code, so switched to that

Reply:
It seems we cannot use `TaskQueryString()` because all the caching logic is based on the `jobQuery` where coordinator evaluation does not happen. 


When we use `TaskQueryString()` at the stage of caching, the parameters have already been resolved and that would be cached. Luckily we have the tests in `coordinator_evaluation_modify` which caught this subtle case.

I think this change still requires a little more attention from this perspective and memory context perspectives.

I added a comment on this

Reply:
Makes sense, though I wish this was a little more obvious from the code.

Reply:
I think an important thing to keep in mind is how this relates to function evaluation.

The reason it is ok to deparse a parameterized query for the local execution case, but not quite ok in the general case is that the insert may have fields like `stable_or_volatile_function($1)` which should in general be evaluated on the coordinator of the query. In the local execution case, we can allow those function calls to be evaluated when the query on the shard is execution, because that's still on the coordinator. That is what makes local execution special.

Reply:
I added a comment based on your comment. I think there is one more confusing thing, which is not related to this PR, but more generic confusion.

As you noted, we rely on Postgres to evaluate functions/params for cached local queries. However, we still do the function/param evaluation for those queries on the **Task** query. That's why, when we do `set citus.log_remote_commands TO ON`, even if the query has param/functions, we see the constant values:
```SQL
execute p2(2);
NOTICE:  executing the command locally: INSERT INTO public.users_table_102012 (user_id, value_1) VALUES (10, 2)
```

Where, the plan that is being executed has the params and functions:
```SQL
            {TARGETENTRY 
            :expr 
               {FUNCEXPR 
               :funcid 25339 
               :funcresulttype 23 
               :funcretset false 
               :funcvariadic false 
               :funcformat 0 
               :funccollid 0 
               :inputcollid 0 
               :args <> 
               :location 65
               }

...
           {TARGETENTRY 
            :expr 
               {PARAM 
               :paramkind 0 
               :paramid 1 
               :paramtype 23 
               :paramtypmod -1 
               :paramcollid 0 
               :location 100
               }
            :resno 3 
            :resname value_1 
            :ressortgroupref 0 
            :resorigtbl 0 
            :resorigcol 0 
            :resjunk false
            }
```


I think we can live with this for now, but would be good to improve separately for better debugging.

Reply:
Actually might be interesting to see the performance penalty of coordination evaluation for local cached plans. Could help to gain some cpu cycles.

Reply:
I think we need to do function/parameter evaluation to do the shard pruning, though we only really need to do it for the distribution column.